### PR TITLE
feat(styled-system): introduce semantic tokens

### PR DESCRIPTION
**Pull request by TimKolberger**
## 📝 Description

Introducing **semantic tokens**

Semantic tokens provide the ability to create css variables which can change
with a CSS condition.

```tsx
extendTheme({
  colors: {
    900: "#171923",
  },
})
;<Text color="gray.900">will always be gray.900</Text>
```

```tsx
extendTheme({
  colors: {
    50: "#F7FAFC",
    900: "#171923",
  },
  semanticTokens: {
    colors: {
      text: {
        default: "gray.900",
        _dark: "gray.50",
      },
    },
  },
})
;<Text color="text">
  will be gray.900 in light mode and gray.50 in dark mode
</Text>
```

```tsx
const theme = extendTheme({
  colors: {
    red: {
      100: "#ff0010",
      400: "#ff0040",
      500: "#ff0050",
      700: "#ff0070",
      800: "#ff0080",
    },
  },
  semanticTokens: {
    colors: {
      error: "red.500", // create a token alias
      success: "red.100",
      primary: {
        // set variable conditionally with pseudo selectors like `_dark` and `_light`
        // use `default` to define fallback value
        default: "red.500",
        _dark: "red.400",
      },
      secondary: {
        default: "red.800",
        _dark: "red.700",
      },
    },
  },
})
```

## 💣 Is this a breaking change (Yes/No):

No

## 📝 Additional Information

* [Please see this codesandbox for a demo](https://codesandbox.io/s/friendly-hooks-k3055?file=/src/index.tsx)
* [Chakra UI Docs PR](https://github.com/chakra-ui/chakra-ui-docs/pull/226)

### CODE REVIEW

File: packages/color-mode/src/color-mode.utils.ts [hunk 1]
```diff
@@ -75,12 +75,13 @@ export function addListener(
...
   get: () =>
-    document.documentElement.style.getPropertyValue(
+    (document.documentElement.style.getPropertyValue(
       "--chakra-ui-color-mode",
-    ) as ColorMode | "",
+    ) || document.documentElement.dataset.theme) as ColorMode | "",
   set: (mode: ColorMode) => {
     if (isBrowser) {
       document.documentElement.style.setProperty("--chakra-ui-color-mode", mode)
+      document.documentElement.setAttribute("data-theme", mode)
```
Comment (TimKolberger) [inline packages/color-mode/src/color-mode.utils.ts:84]: I am syncing the current color mode to the document root: e.g. `<html data-theme="dark">`.
This allows the `_dark` pseudo selector (`".chakra-ui-dark &, [data-theme=dark] &, &[data-theme=dark]"`) to pick up the current color mode.

File: .changeset/dull-keys-occur.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        _dark: "gray.50",
+      },
+    },
+  },
+})
+
+const App = () => (
+  <ChakraProvider theme={customTheme}>
+    <Text color="text">
+      will be gray.900 in light mode and gray.900 in dark mode
```
Comment (TimKolberger) [inline .changeset/dull-keys-occur.md:49]:
```suggestion
      will be gray.900 in light mode and gray.50 in dark mode
```

File: packages/color-mode/src/color-mode.utils.ts (hunk 1 above)
Comment (TimKolberger) [reply packages/color-mode/src/color-mode.utils.ts:89]: As discussed with Sage we are going to deprecate the internal CSS var `--chakra-ui-color-mode` with the next major release.

**Review by segunadebayo — APPROVED** [review]
Looks good to me!

**Comment by tiagooliveira08** [comment]
good, i need this!

**Comment by tiagooliveira08** [comment]
work with backgroundColor or borderColor?

**Comment by TimKolberger** [comment]
A semantic token can be used in a CSS prop like a normal token.

A semantic token in the colors scale can be used in backgroundColor, borderColer etc.

A semantic token in the fonts scale can be used in fontFamily etc.

Please note: when using _dark or _light selector in a semantic token, check that it will not be overridden by component styles.

**Pull request merged**